File: 0036-valid-sudoku/0036-valid-sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        # approach: check rows, then cols, then sub-boxes
        
        def checkBox(row_start, col_start):   
            hashset = set()

            for i in range(row_start, row_start + 3):
                for j in range(col_start, col_start + 3):
                    n = board[i][j]

                    if n != '.':
                        if n in hashset:
                            return False
                        
                        hashset.add(n)

            return True
        
        row_hash = collections.defaultdict(set)
        col_hash = collections.defaultdict(set)
        
        # check rows and cols
        for i in range(len(board)):
            for j in range(len(board)):
                r = board[i][j]
                c = board[j][i]
                
                if (r != '.'):
                    if r in row_hash[i]:
                        return False
                    
                    row_hash[i].add(r)
                
                if (c != '.'):
                    if c in col_hash[i]:
                        return False
                    
                    col_hash[i].add(c)
        
        # check sub-boxes
        for i in range(0, 9, 3):
            for j in range(0, 9, 3):
                if not checkBox(i, j):
                    return False
        
        return True
                

File: 0036-valid-sudoku/0036-valid-sudoku.java
class Solution {
    public boolean isValidSudoku(char[][] board) {
        // approach: hashset of seen nums, check rows, then cols, the sub-boxes
        
        int size = board[0].length;
        
        // check rows and cols
        for (int i = 0; i < size; i++) {            
            Set<Character> rowSet = new HashSet<>();
            Set<Character> colSet = new HashSet<>();

            for (int j = 0; j < size; j++) {
                char r = board[i][j];
                char c = board[j][i];
                
                if (r != '.') {
                    if (rowSet.contains(r)) {
                        return false;
                    }

                    rowSet.add(r);
                }
                
                if (c != '.') {
                    if (colSet.contains(c)) {
                        return false;
                    }

                    colSet.add(c);
                }
            }
        }
        
        // check boxes
        int startI = 3;
        int startJ = 0;
        
        for (int i = 0; i < 9; i += 3) {
            for (int j = 0; j < 9; j += 3) {
                if (!checkBox(board, i, j)) {
                    return false;
                }
            }
        }
        
        return true;
    }
    
    private boolean checkBox(char[][] board, int startI, int startJ) {
        Set<Character> boxSet = new HashSet<>();
        
        for (int i = startI; i < startI + 3; i++) {
            for (int j = startJ; j < startJ + 3; j++) {
                char n = board[i][j];
                
                System.out.println(board[i][j]);
                
                if (n != '.') {
                    if (boxSet.contains(n)) {
                        return false;
                    }
                    
                    boxSet.add(n);
                }
            }
        }
        
        return true;
    }
}